blockprod/src/detail/mod.rs

```diff
@@ -47,158 +85,291 @@ impl BlockProduction {
         chainstate_handle: ChainstateHandle,
         mempool_handle: MempoolHandle,
         time_getter: TimeGetter,
-        builder_tx: mpsc::UnboundedSender<BlockBuilderControlCommand>,
+        mining_thread_pool: Arc<slave_pool::ThreadPool>,
     ) -> Result<Self, BlockProductionError> {
         let block_production = Self {
             chain_config,
             chainstate_handle,
             mempool_handle,
             time_getter,
-            builder_tx,
+            all_jobs: BTreeMap::new(),
+            mining_thread_pool,
         };
         Ok(block_production)
     }
-}
 
-#[async_trait::async_trait]
-impl BlockProductionInterface for BlockProduction {
-    fn stop(&self) -> Result<(), BlockProductionError> {
-        self.builder_tx
-            .send(BlockBuilderControlCommand::Stop)
-            .map_err(|_| BlockProductionError::BlockBuilderChannelClosed)?;
-        Ok(())
+    pub fn time_getter(&self) -> &TimeGetter {
+        &self.time_getter
+    }
+
+    pub fn stop_all_jobs(&mut self) {
+        let mut all_jobs = Vec::new();
+        while let Some((key, handle)) = self.all_jobs.pop_first() {
+            all_jobs.push((key, handle));
+        }
+
+        log::info!("Cancelling {} jobs", all_jobs.len());
+
+        for (key, handle) in all_jobs.drain(..) {
```

Comment:
Unless I'm mistaken, the `pop_first()` call in the above `while()` loop will also drain the hashmap, so there's no need to recreate it in order to drain.